File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <cmath>
using namespace std;


int get_stayed_time(map<string, string> cars, string carNum, string out) {
    string in = cars[carNum];
    int in_time = stoi(in.substr(0, 2)) * 60 + stoi(in.substr(3, 2));
    int out_time = stoi(out.substr(0, 2)) * 60 + stoi(out.substr(3, 2));
    return out_time - in_time;
}

int get_price(int stayed_time, vector<int> fees) {
    if (stayed_time <= fees[0]) return fees[1];
    double c = double(stayed_time - fees[0]) / fees[2];
    return fees[1] + ceil(c) * fees[3];
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    map<string, int> stayed_time;
    map<string, string> cars;
    string num, time, act;
    for (int i = 0; i < records.size(); i++) {
        stringstream ss(records[i]);
        ss >> time >> num >> act;
        if (act == "IN") {
            cars[num] = time;
        } else if (act == "OUT") {
            stayed_time[num] += get_stayed_time(cars, num, time);
            cars.erase(num);
        }
    }
    for (auto it = cars.begin(); it != cars.end(); it++)
        stayed_time[it->first] += get_stayed_time(cars, it->first, "23:59");

    for (auto it = stayed_time.begin(); it != stayed_time.end(); it++)
        answer.push_back(get_price(it->second, fees));

    return answer;
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> map = new HashMap<>();
        Map<String, Integer> feeMap = new TreeMap<>();
        
        for (String record : records) {
            String[] split = record.split(" ");
        
            int converted = convert(split[0]);
            
            if (split[2].equals("IN")) {
                map.put(split[1], converted);
            } else {
                int inTime = map.get(split[1]);
                int outTime = convert(split[0]);
            
                feeMap.put(split[1], feeMap.getOrDefault(split[1], 0) + outTime - inTime);
                map.remove(split[1]);
            }
            
        }
        
        for (String no : map.keySet()) {
            int outTime = convert("23:59");
            int inTime = map.get(no);
            
            feeMap.put(no, feeMap.getOrDefault(no, 0) + outTime - inTime);
        }
        
        int[] answer = new int[feeMap.size()];
            
        int idx = 0;
        for (Integer time : feeMap.values()) {
            if (time <= fees[0]) {
                answer[idx++] = fees[1];
            } else {
                int val = fees[1] + (int) Math.ceil((double)(time - fees[0]) / fees[2]) * fees[3];
                answer[idx++] = val;
            }
        }
        
        return answer;
    }
    
    public int convert(String time) {
        String[] split = time.split(":");
        return Integer.parseInt(split[0]) * 60 + Integer.parseInt(split[1]);
    }
}